torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -420,13 +420,17 @@ const auto kLoopbackAddress = "127.0.0.1";
 
 // static
 void ProcessGroupGloo::AsyncWork::execute(c10::intrusive_ptr<AsyncWork> work) {
-  std::exception_ptr eptr;
   try {
     work->run();
   } catch (...) {
-    eptr = std::current_exception();
+    work->finishWorkGlooError(std::current_exception());
+    return;
   }
-  work->finish(eptr);
+
+  // FIXME: We need to call it here since Future completion requires all
+  // the work to be synchronized to CUDA.
```

Comment:
Since we now have CUDA support in `ivalue::Future`, should this synchronization work be done by `ivalue::Future` instead?  cc @lw

Reply:
completely agree. 

BTW, we used to do some other things in synchronize() (e.g. copying tensors), however seems like now all derived versions of synchronize() match ivalue::Future::synchronizeWithCurrentStreams()

The plan is to deprecate work API in 1.10, so we don't have any more versions of synchronize() and then use
ivalue::Future::synchronizeWithCurrentStreams() directly.

cc @lw